storage/src/main/java/tech/pegasys/artemis/storage/V2MapDatabase.java

```diff
@@ -0,0 +1,291 @@
+  }
+
+  public static Database createInMemory() {
+    return new V2MapDatabase(DBMaker.memoryDB());
+  }
+
+  private V2MapDatabase(final Maker dbMaker) {
+    db = dbMaker.transactionEnable().make();
+    time = db.atomicVar("time", new UnsignedLongSerializer()).createOrOpen();
+    genesisTime = db.atomicVar("genesisTime", new UnsignedLongSerializer()).createOrOpen();
+    justifiedCheckpoint =
+        db.atomicVar("justifiedCheckpoint", new MapDBSerializer<>(Checkpoint.class)).createOrOpen();
+    bestJustifiedCheckpoint =
+        db.atomicVar("bestJustifiedCheckpoint", new MapDBSerializer<>(Checkpoint.class))
+            .createOrOpen();
+    finalizedCheckpoint =
+        db.atomicVar("finalizedCheckpoint", new MapDBSerializer<>(Checkpoint.class)).createOrOpen();
+
+    finalizedRootsBySlot =
+        db.treeMap("finalizedRootsBySlot", new UnsignedLongSerializer(), new Bytes32Serializer())
+            .createOrOpen();
+    finalizedBlocksByRoot =
+        db.hashMap(
+                "finalizedBlocksByRoot",
+                new Bytes32Serializer(),
+                new MapDBSerializer<>(BeaconBlock.class))
+            .createOrOpen();
+    finalizedStatesByRoot =
+        db.hashMap(
+                "finalizedStatsByRoot",
+                new Bytes32Serializer(),
+                new MapDBSerializer<>(BeaconState.class))
+            .createOrOpen();
+
+    hotBlocksByRoot =
+        db.hashMap(
+                "hotBlocksByRoot",
+                new Bytes32Serializer(),
+                new MapDBSerializer<>(BeaconBlock.class))
+            .createOrOpen();
+    hotStatesByRoot =
+        db.hashMap(
+                "hotStatesByRoot",
+                new Bytes32Serializer(),
+                new MapDBSerializer<>(BeaconState.class))
+            .createOrOpen();
+
+    checkpointStates =
+        db.hashMap(
+                "checkpointStates",
+                new MapDBSerializer<>(Checkpoint.class),
+                new MapDBSerializer<>(BeaconState.class))
+            .createOrOpen();
+
+    latestMessages =
+        db.hashMap(
+                "latestMessages",
+                new UnsignedLongSerializer(),
+                new MapDBSerializer<>(Checkpoint.class))
+            .createOrOpen();
+
+    // Recreate hotRootsBySlotCache
+    hotBlocksByRoot.forEach(this::addToHotRootsBySlotCache);
+  }
+
+  @Override
+  public void storeGenesis(final Store store) {
+    time.set(store.getTime());
+    genesisTime.set(store.getGenesisTime());
+    justifiedCheckpoint.set(store.getJustifiedCheckpoint());
+    finalizedCheckpoint.set(store.getFinalizedCheckpoint());
+    bestJustifiedCheckpoint.set(store.getBestJustifiedCheckpoint());
+    store
+        .getBlockRoots()
+        .forEach(
+            root -> {
+              addHotBlock(root, store.getBlock(root));
+              hotStatesByRoot.put(root, store.getBlockState(root));
+            });
+    checkpointStates.put(
```

Comment:
Do we need to store the state associated with `bestJustifiedCheckpoint`?

Reply:
Worth doing for good measure.  At genesis the best justified state is the same as the justified state but definitely good to avoid coupling the two pieces of code more than we need to.